### PR TITLE
Refactor reorderable list semantics

### LINKED ISSUE #71396: ReorderableListView merges semantics of list item's children preventing activation of child widget
Internal: b/273502565

## Steps to Reproduce

1. Run `flutter create bug`.
2. Update main.dart as follows: 
``` dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: ReorderableListView(
          children: [
            _createItem('Item 1'),
            _createItem('Item 2'),
          ],
          onReorder: (from, to) {},
        ),
      ),
    );
  }

  ListTile _createItem(String title) {
    return ListTile(
      key: UniqueKey(),
      title: Text(title),
      trailing: PopupMenuButton<String>(
        onSelected: (String result) {
          print('selected: $result');
        },
        itemBuilder: (BuildContext context) => <PopupMenuEntry<String>>[
          PopupMenuItem<String>(
            value: 'rename',
            child: Text('Rename'),
          ),
          PopupMenuItem<String>(
            value: 'delete',
            child: Text('Delete'),
          ),
        ],
      ),
    );
  }
}
```
3. ... When enabling TalkBack/Voiceover you cannot activate the PopupMenuButton

**Expected results:** 

You should be able to active PopupMenuButton

**Actual results:** 

As Semantics are merged for list item's children PopupMenuButton cannot be activated.

<details>
  <summary>Logs</summary>

```
```

```
```

```
[✓] Flutter (Channel stable, 1.22.4, on Linux, locale es_ES.UTF-8)
    • Flutter version 1.22.4 at /home/angel/flutter/soft/flutter
    • Framework revision 1aafb3a8b9 (hace 2 semanas), 2020-11-13 09:59:28 -0800
    • Engine revision 2c956a31c0
    • Dart version 2.10.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /home/angel/android/soft/sdk/
    • Platform android-30, build-tools 30.0.0
    • ANDROID_HOME = /home/angel/android/soft/sdk/
    • ANDROID_SDK_ROOT = /home/angel/android/soft/sdk/
    • Java binary at: /home/angel/android/soft/android-studio/jre/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Android Studio (version 4.0)
    • Android Studio at /home/angel/android/soft/android-studio
    • Flutter plugin version 50.0.1
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.51.1)
    • VS Code at /usr/share/code
    • Flutter extension version 3.16.0

[✓] Connected device (1 available)
    • Pixel 3a (mobile) • 93JAY0BMB2 • android-arm64 • Android 11 (API 30)

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by chunhtai**
fixes https://github.com/flutter/flutter/issues/71396

and move semantics to sliverReoderableList

all i18n related change is in first commit.

reorderable list related change is in second commit

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/reorderable_list.dart
Comment (chunhtai) [inline packages/flutter/lib/src/widgets/reorderable_list.dart:445]: I make this required because this has to be the same as the scrollview, so It feels weird to me that we let people omit this property.

**Comment by chunhtai** [comment]
I have separated out i18n and reorderablelist changes into separate commit, hopefully this makes it easier to review

**Comment by chunhtai** [comment]
a friendly bump

File: packages/flutter/lib/src/widgets/reorderable_list.dart
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/reorderable_list.dart:448]: It's a bit surprising that you have to tell the sliver in which direction it scrolls (and keep that in sync manually with e.g. the ScrollView). I don't think we do this for any other sliver. Can this sliver not find out the scroll direction by e.g. looking up `Scrollable.of`?

cc @Piinks

File: dev/tools/localization/bin/gen_localizations.dart
Comment (goderbauer) [inline dev/tools/localization/bin/gen_localizations.dart:589]: Should this be widgetsLocaleToResources? Maybe add a test that would have failed for this case

File: dev/tools/localization/bin/gen_localizations.dart
Comment (goderbauer) [inline dev/tools/localization/bin/gen_localizations.dart:72]: wondering: why are they not all const?

**Review by goderbauer — COMMENTED** [review]
The first commit is super-hard to review. Maybe separate hand-crafted changes to the tool and automatically generated changes into two separate commits. 

Also, do the changes to the tool have to happen in the same PR that modifies the semantics of the reorderable list? Should/could those land separately?

Last, but not least, please get someone more familiar with `dev/tools/localization` and `flutter_localizations` to review those bits.

File: packages/flutter/lib/src/material/reorderable_list.dart
Comment (Piinks) [inline packages/flutter/lib/src/material/reorderable_list.dart:423]: What if this conflicts with the scroll direction provided by the constraints during layout?

File: packages/flutter/lib/src/widgets/reorderable_list.dart
Comment (Piinks) [reply packages/flutter/lib/src/widgets/reorderable_list.dart:448]: Actually, it looks like the scroll direction is already known, there is a _scrollDirection available in the widgets/SliverReorderableListState

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/reorderable_list.dart:448]: Rendering layer may be too late to apply the semantics actions to children.

Comment (Piinks) [reply packages/flutter/lib/src/widgets/reorderable_list.dart:448]: Right right, SliverReorderableList does not actually have its own rendering object anyways, my mistake. Does the _scrollDirection that is already available from the state work?

File: dev/tools/localization/bin/gen_localizations.dart
Comment (chunhtai) [reply dev/tools/localization/bin/gen_localizations.dart:72]: this bool decide whether to add `const` keyword when calling factory. Since material/cupertino pass in parameter to when calling factory, they can't use the `const`. Will update the name to be more precise

File: packages/flutter/lib/src/widgets/reorderable_list.dart
Comment (chunhtai) [reply packages/flutter/lib/src/widgets/reorderable_list.dart:448]: @Piinks Do you think using Scrollable.maybeOf would be a good solution?

Comment (chunhtai) [reply packages/flutter/lib/src/widgets/reorderable_list.dart:448]: oh yeah, I missed it, it looks like it was using scrollable.of. I will just use that.

**Comment by chunhtai** [comment]
turning this into draft until https://github.com/flutter/flutter/pull/123620/files is merged

File: packages/flutter/lib/src/material/material_localizations.dart
Comment (chunhtai) [reply packages/flutter/lib/src/material/material_localizations.dart:427]: If someone is using it, then yes. It is hard to imagine how people will reuse this though.

I am not sure whether people would extend this class. @thkim1011, do you know if there is use case for that?

Comment (chunhtai) [reply packages/flutter/lib/src/material/material_localizations.dart:427]: This is technically not a breaking change given all tests pass. The breaking change guide states it would be optional whether we announce this as a courtesy. I personally think it is very unlikely this would be a breaking change. WDYT?

Comment (Piinks) [reply packages/flutter/lib/src/material/material_localizations.dart:427]: If no tests broke that sounds fine to me. I would probably do a global test run in Google before this lands though to make sure.

**Review by Piinks — APPROVED** [review]
Flutter_LGTM

**Review by goderbauer — APPROVED** [review]
LGTM!

**Comment by borjandev** [comment]
@Piinks @chunhtai This breaks https://pub.dev/packages/reorderables

```
Could not build the precompiled application for the device.
Error (Xcode): ../../.pub-cache/hosted/pub.dev/reorderables-0.6.0/lib/src/widgets/reorderable_wrap.dart:737:34: Error: The getter 'reorderItemToStart' isn't defined for the class 'MaterialLocalizations'.
```

However, maybe the resolution is for the package maintainer, I believe @hanshengchiu to update the package instead

**Comment by chunhtai** [comment]
I will prepare a migration guide and dart fix, it looks like it is more widely used than I think

**Comment by Piinks** [comment]
Thanks @chunhtai !

@borjandev unfortunately any change could potentially break a package on pub.dev 
We cannot account for all of them, but we do have several signals we use to indicate breaking changes to us.

Any contributor can add tests to [`flutter/tests`](https://github.com/flutter/tests), and if a change breaks them, we won't make the change without working with the contributor to ensure the package is not broken. Maybe https://pub.dev/packages/reorderables would like to contribute tests.

We also look at how our own tests are affected by a change, and how customer tests inside of Google are affected.

In this case, our signals worked and the change was reverted after we found Google tests were broken by the change. 

To learn more about our breaking change policy and how we determine breaking changes, you can read [the breaking change section in the wiki](https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes).

**Pull request merged**
